pkg/kubelet/container/helpers.go

```diff
@@ -130,9 +131,32 @@ func ExpandContainerCommandOnlyStatic(containerCommand []string, envs []v1.EnvVa
 	return command
 }
 
-func ExpandContainerVolumeMounts(mount v1.VolumeMount, envs []EnvVar) (expandedSubpath string) {
-	mapping := expansion.MappingFuncFor(EnvVarsToMap(envs))
-	return expansion.Expand(mount.SubPath, mapping)
+func ExpandContainerVolumeMounts(mount v1.VolumeMount, envs []EnvVar) (expandedSubpath string, ok bool) {
+
```

Comment:
rather than doing regex parsing ourselves, would it be simpler to track missing keys in the mapping function, like this?

```
func ExpandContainerVolumeMounts(mount v1.VolumeMount, envs []EnvVar) (string, error) {
  envmap := EnvVarsToMap(envs)
  missingKeys := sets.NewString()
  expanded := expansion.Expand(mount.SubPathExpr, func(key string) string{
    value, ok := envmap[input]
    if !ok || len(value) == 0 {
      missingKeys.Insert(key)
    }
    return value
  })
  if len(missingKeys) > 0 {
    return "", fmt.Errorf("missing value for %s",strings.Join(missingKeys.List(), ", "))
  }
  return expanded, nil
}
```

Reply:
that would also let us return the missing envvars in the error, which is nice

Reply:
Not sure I quite get this routine

Reply:
Not sure I understand this construction `value, ok := envmap[input]` - what does `input` refer to?

Reply:
sorry... changed my variable name halfway through writing it... I mean `key`

Reply:
It also doesn't seem to fulfil all the test cases which include `foo/$(ODD_NAME)/$(POD_NAME)` which seems to just return POD_NAME on first testing of this code

Reply:
And this one doesn't behave either `POD_NAME$(ANNOTATION)` - this is why I use the regex to split all the cases at the boundaries of the envvars

Reply:
expansion.Expand already splits tokens for you, we shouldn't need to reimplement that...

your mapping function determines what values to replace the specified keys with, and can record missing values, which cause ExpandContainerVolumeMounts to return an error

Reply:
Unless I am missing something I cant see this handle the case when the subpath expressruon contains multiple values which might not be in the environment variable list

Reply:
the unit test should look like this:

```
	for _, tc := range cases {
		actualSubPath, err := ExpandContainerVolumeMounts(tc.container.VolumeMounts[0], tc.envs)
		ok := err == nil
		if e, a := tc.expectedOk, ok; !reflect.DeepEqual(e, a) {
			t.Errorf("%v: unexpected validation failure of subpath; expected %v, got %v", tc.name, e, a)
		}
		if !ok {
			// if ExpandContainerVolumeMounts returns an error, we don't care what the actualSubPath value is
			continue
		}
		if e, a := tc.expectedSubPath, actualSubPath; !reflect.DeepEqual(e, a) {
			t.Errorf("%v: unexpected subpath; expected %v, got %v", tc.name, e, a)
		}
		if e, a := tc.expectedMountPath, tc.container.VolumeMounts[0].MountPath; !reflect.DeepEqual(e, a) {
			t.Errorf("%v: unexpected mountpath; expected %v, got %v", tc.name, e, a)
		}
	}
```

Reply:
that verifies that all the `expectedOk:true` cases pass successfully, and all the `expectedOk:false` cases get an error as expected.

when we get an error, we don't care what the resolved value is:

```
			subPath, err = kubecontainer.ExpandContainerVolumeMounts(mount, expandEnvs)

			if err != nil {
				return nil, cleanupAction, fmt.Errorf("volume subpath expansion failed: %v", err)
			}
```

Reply:
Sorry, I wasnt clear. I have made all the test cases run but there are some that wont work with this code suggestion. So if the subpath is set to something like $(ODD_NAME)/$(POD_NAME) and ODD_NAME is not defined in the env list then it wont get parsed

Reply:
> if the subpath is set to something like $(ODD_NAME)/$(POD_NAME) and ODD_NAME is not defined in the env list then it wont get parsed

is this the test case you mean?
```
{
	name: "subpaths with missing env variables",
	container: &v1.Container{
		VolumeMounts: []v1.VolumeMount{{SubPathExpr: "$(ODD_NAME)/$(POD_NAME)"}},
	},
	envs: []EnvVar{
		{
			Name:  "POD_NAME",
			Value: "bar",
		},
	},
	expectedSubPath:   "",
	expectedMountPath: "",
	expectedOk:        false,
},
```

that fails as I'd expect it to

Reply:
Thanks. I revisited the test cases and checked the results. These do seem okay, I misinterpreted the error condition being returned in one of the cases. I have pushed this change up